### PR TITLE
Add real time collaboration coding editor

### LINKED ISSUE #82: Implement real time collaboration coding editor

### PR DESCRIPTION
**Pull request by ys112**
[Summarised by copilot]

These changes collectively establish a new collaboration service and integrate it with the frontend, enabling real-time collaborative coding features.

### Frontend Changes:
* Added a new WebSocket URL for the collaboration service in the `.env` file. (`frontend/.env`)
* Updated `package.json` to include new dependencies required for the collaboration feature, such as `@codemirror` and `yjs`. (`frontend/package.json`) [[1]](diffhunk://#diff-da6498268e99511d9ba0df3c13e439d10556a812881c9d03955b2ef7c6c1c655R28-R31) [[2]](diffhunk://#diff-da6498268e99511d9ba0df3c13e439d10556a812881c9d03955b2ef7c6c1c655R42-R48)
* Introduced a new `CodingEditor` component that sets up a collaborative coding environment using `yjs` and `codemirror`. (`frontend/src/components/Collaboration/CodingEditor.tsx`)
* Modified the collaboration route to include the new `CodingEditor` component. (`frontend/src/routes/_authenticated/collaboration.tsx`)

### Backend Changes:
* Implemented document retrieval and storage functions in the collaboration service. (`services/collaboration/src/controller/doc-controller.ts`)
* Added user auth function for the collaboration service. (`services/collaboration/src/utils/logger.ts`, `services/collaboration/src/utils/verifyToken.ts`) [[1]](diffhunk://#diff-4e90508c2598f915b199ca8470020872ffbfde25bc01376f37b18820337887dfR1-R79) [[2]](diffhunk://#diff-ee1b7cc10d18729e6ec7120a1653f6405c71cf1a733b3e0fedb616139b2e8b5eR1-R39)

### CODE REVIEW

File: services/collaboration/.env.development.example
Comment (Cloud7050) [inline services/collaboration/.env.development.example:1]: This file does not seem needed as we are checking in `.env.development`, and it is the same

File: services/collaboration/.env.development
Comment (Cloud7050) [inline services/collaboration/.env.development:1]: Should we rename this to `.env`? Also, the env vars don't seem to be used in the code at least for now.

File: services/collaboration/src/utils/verifyToken.ts
Comment (Cloud7050) [inline services/collaboration/src/utils/verifyToken.ts:1]: Nice to have: Dedupe some of the commonly reused logic into `@common` (#40)

File: services/collaboration/README.md
Comment (Cloud7050) [inline services/collaboration/README.md:1]: We could file an issue as a TODO for these docs

File: services/collaboration/.env.development
Comment (ys112) [reply services/collaboration/.env.development:1]: Oh yes, will change to .env. The env is currently used for firebase and port and user service api, more will come when we integrate the other services.

File: services/collaboration/.env.development.example
Comment (ys112) [reply services/collaboration/.env.development.example:1]: Ok, will remove.

File: services/collaboration/README.md
Comment (ys112) [reply services/collaboration/README.md:1]: Thanks for spotting, will update the doc.

File: services/collaboration/README.md [hunk 1]
```diff
@@ -1,25 +1,18 @@
...
-## Scripts
-Use `npm run <script>` or `yarn <script>` to run available scripts in `package.json`.
+## Development Instructions
 
-**Window users**: If you encounter this error: `'NODE_ENV' is not recognized as an internal or external command, operable program or batch file.`, add `cross-env` before every `NODE_ENV`. The `dev` and `test` scripts have been updated to use `cross-env`.
+Unless otherwise specified, commands should be run in the service root.
 
-- `dev`: Start the server in development mode, with hot reloading (`ts-node-dev`) enabled
+### Running the service (development) 
+`pnpm run dev:collaboration` : Run the dev script for this collaboration service.
```
Comment (Cloud7050) [inline services/collaboration/README.md]: This one seems like it'll only work from the monorepo root. Maybe `pnpm dev` is enough from service root.

**Review by Cloud7050 — APPROVED** [review]

**Review by asaierika — APPROVED** [review]

**Pull request merged**